### PR TITLE
feat(canary): Add test to check query results with and without cache.

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:
Recently, we added support for `Cache-Control` header for loki. 

This PR uses it to test if same query for same data returns same results with and without results cache.

Easy to catch any bug in results cache if there are any differences.

NOTE: This currently does `count_over_time` instant query for the cache test. In future we can extend it for range queries as well.

This PR introduces to additional metrics for `loki-canary` to track any differences in this test.
```
loki_canary_cache_test_query_results_total{status=~"success|fail"}
loki_canary_cache_test_query_results_diff_total
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

This is one of the step to gain more confidence on recently added "results cache" support for instant query.

Manually ingested logs with older timestamps to create discrepancy with results cache. And counters seems to be working fine.

```
# HELP loki_canary_cache_test_query_results_diff_total counts number of times the query results was different with and without cache 
# TYPE loki_canary_cache_test_query_results_diff_total counter
loki_canary_cache_test_query_results_diff_total 16
# HELP loki_canary_cache_test_query_results_total counts number of times the query results test requests are done 
# TYPE loki_canary_cache_test_query_results_total counter
loki_canary_cache_test_query_results_total{status="success"} 23
```

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/canary/comparator/comparator_test.go [hunk 1]
```diff
@@ -360,6 +360,43 @@ func TestSpotCheck(t *testing.T) {
...
+	mr.countOverTime = 2.3               // value not important
+	mr.noCacheCountOvertime = 2.30000005 // different than `countOverTime` value but withing tolerance
+	c.cacheTest(now)
+	assert.Equal(t, 0, queryResultsDiff.(*mockCounter).count)
+
+	// This avoids a panic on subsequent test execution,
+	// seems ugly but was easy, and multiple instantiations
+	// of the comparator should be an error
+	prometheus.Unregister(responseLatency)
+
```
Comment (MichelHollands) [inline pkg/canary/comparator/comparator_test.go:397]: Empty line can be removed

**Review by MichelHollands — APPROVED** [review]
Some cosmetic changes required but otherwise fine

**Comment by kavirajk** [comment]
Thanks @MichelHollands addressed your comments.

**Comment by kavirajk** [comment]
Checked with internal dev loki clusters. Manually ingested some logs with old timestamp to create the diff.

Screenshot 2024-06-03 at 17 54 31

**Pull request merged**